### PR TITLE
Add build target for Java 11

### LINKED ISSUE #3232: Set up target for JDK 11
JDK 11 releases September 25, 2018.  We should test and verify.

### PR DESCRIPTION
**Pull request by briandealwis**
This patch creates a new target on Travis for testing with Java 11.

Although Tycho does not yet support compiling for Java 11, it doesn't much matter since we're supporting Java 8.  We use Tycho's toolchains support to compile for Java 8 but run with Java 11, similar to when we still supported Eclipse Mars / 4.5 with Java 7/Java 8.

  - The process has been significantly simplified by using Maven's `jdk` profile activation based on the JVM used for the build.
  - I reworked the `.travis.yml` build matrix to add nice job names screen shot 2018-09-07 at 4 57 28 pm
  - bumped Tycho to 1.2.0
  - bumped maven-compiler-plugin and errorprone to handle Java 11 class files for the JRE libraries

I'm a bit puzzled that I didn't need to add `--add-modules=ALL-SYSTEM` for tycho-surefire launching.

Fixes #3232

### CODE REVIEW

**Comment by chanseokoh** [comment]
Tests are failing, unfortunately.

File: .travis.yml [hunk 1]
```diff
@@ -40,7 +47,6 @@ script: ./mvnw -V -B --fail-at-end verify ${MAVEN_FLAGS}
      -Ptravis -Perrorprone
      ${ECLIPSE_TARGET:+-Declipse.target=${ECLIPSE_TARGET}}
      --toolchains=.travisci/toolchains.xml
-     -Dtoolchain.java.runtime=${JAVA_RUNTIME}
```
[the remark this answers is not in the document]
Comment (briandealwis) [reply .travis.yml]: Exactly!

File: .travisci/toolchains.xml [hunk 2]
```diff
@@ -12,4 +28,15 @@
       <jdkHome>/usr/lib/jvm/java-8-oracle/jre</jdkHome>
     </configuration>
   </toolchain>
+  <toolchain>
+    <type>jdk</type>
+    <provides>
+      <id>JavaSE-1.7</id>
```
[the remark this answers is not in the document]
Comment (briandealwis) [reply .travisci/toolchains.xml]: We have a fragment  of `org.eclipse.jst.server.core` (in `third_party/com.google.cloud.tools.eclipse.jst.server.core`) which is JavaSE-1.7.

**Comment by briandealwis** [comment]
The Java 8 builds are failiing because the `toolchains.xml` references the jdk11 location but it isn't actually installed.
> 2462931 [ERROR] Failed to execute goal org.eclipse.tycho.extras:tycho-eclipserun-plugin:1.2.0:eclipse-run (default) on project verify-feature-completeness: Error while executing platform: Non-existing JDK home configuration at /home/travis/openjdk11

The Java 11 build is failing because:
```
247204 [INFO] -------------------------------------------------------------
247204 [ERROR] COMPILATION ERROR : 
247204 [INFO] -------------------------------------------------------------
247204 [ERROR] /home/travis/build/GoogleCloudPlatform/google-cloud-eclipse/plugins/com.google.cloud.tools.eclipse.test.dependencies/src/com/google/cloud/tools/eclipse/test/dependencies/LogConfigurator.java:[27,8] cannot access java.lang.Object
  bad class file: /modules/java.base/java/lang/Object.class
    class file has wrong version 55.0, should be 53.0
    Please remove or make sure it appears in the correct subdirectory of the classpath.
```
which is odd as bumping `maven-compiler-plugin` to 3.8.0 fixed that for me locally.

**Comment by briandealwis** [comment]
Oh yeah, that makes sense.

```
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.22 sec <<< FAILURE! - in com.google.cloud.tools.eclipse.bugreport.ui.BugReportCommandHandlerTest
testFormatReportUrl(com.google.cloud.tools.eclipse.bugreport.ui.BugReportCommandHandlerTest)  Time elapsed: 0.197 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: (a string starting with "1.7." or a string starting with "1.8.")
     but: was "11"
	at com.google.cloud.tools.eclipse.bugreport.ui.BugReportCommandHandlerTest.testFormatReportUrl(BugReportCommandHandlerTest.java:69)
```

Still need to figure out this `eclipse-run` error.

**Comment by briandealwis** [comment]
[JDK11 builds](https://api.travis-ci.org/v3/job/426509286/log.txt) are failing in `ImportMavenAppEngineStandardProjectTest#runImport()`.  
```
03:39:41.123 [Worker-7: Importing Maven projects] WARN  Sisu - Error injecting: org.codehaus.plexus.archiver.jar.JarArchiver
java.lang.ExceptionInInitializerError: null
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:na]
[...]
	at org.eclipse.m2e.core.internal.project.ProjectConfigurationManager.importProjects(ProjectConfigurationManager.java:143) ~[na:na]
[...]
	at org.eclipse.m2e.core.ui.internal.wizards.ImportMavenProjectsJob.runInWorkspace(ImportMavenProjectsJob.java:77) ~[na:na]
[...]
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
	at org.codehaus.plexus.archiver.zip.AbstractZipArchiver.<clinit>(AbstractZipArchiver.java:116) ~[plexus-archiver-2.9.jar:na]

runImport(com.google.cloud.tools.eclipse.integration.appengine.ImportMavenAppEngineStandardProjectTest)  Time elapsed: 4.962 sec  <<< FAILURE!
java.lang.AssertionError: Project does not have AES facet
	at com.google.cloud.tools.eclipse.integration.appengine.ImportMavenAppEngineStandardProjectTest.runImport(ImportMavenAppEngineStandardProjectTest.java:74)
```

It turns out that this is [a bug in the plexus-archiver component that was discovered and fixed for Java 9 in 2015](https://github.com/codehaus-plexus/plexus-archiver/pull/12). Unfortunately [there's nothing we can do to fix it except bump the maven-jar-plugin in our test poms](https://bugs.eclipse.org/bugs/show_bug.cgi?id=516887).

**Comment by briandealwis** [comment]
Java 11 on 2018-09 still failing because it's pulling in the older plexus-archiver.  Which is odd as I can import it into my local IDE (2018-09 on Java 11).
```
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
	at org.codehaus.plexus.archiver.zip.AbstractZipArchiver.<clinit>(AbstractZipArchiver.java:116) ~[plexus-archiver-2.9.jar:na]
	... 88 common frames omitted
```

**Comment by briandealwis** [comment]
I can't reproduce on my machine.  Trying to see if I can reproduce on the Travis docker images.

**Comment by briandealwis** [comment]
Managed to reproduce.  Aha!  The problem is that our `springboot-appengine-standard.zip` project is using `maven-war-plugin:2.6` which pulls in [`plexus-archiver:2.6`](https://github.com/apache/maven-war-plugin/blob/maven-war-plugin-2.6/pom.xml#L131).

**Comment by briandealwis** [comment]
Successful builds! 🎆

**Comment by chanseokoh** [comment]
Awesome!

**Comment by briandealwis** [comment]
Need to tweak the [configuration to enable error-prone](http://errorprone.info/docs/installation).

**Comment by briandealwis** [comment]
PTAL

**Pull request merged**